sky/unit/test/fn3/parent_data_test.dart

```diff
@@ -0,0 +1,287 @@
+import 'package:sky/rendering.dart';
+import 'package:sky/src/fn3.dart';
+import 'package:test/test.dart';
+
+import 'test_widgets.dart';
+import 'widget_tester.dart';
+
+class TestParentData {
+  TestParentData({ this.top, this.right, this.bottom, this.left });
+
+  final double top;
+  final double right;
+  final double bottom;
+  final double left;
+}
+
+void checkTree(WidgetTester tester, List<TestParentData> expectedParentData) {
+  MultiChildRenderObjectElement element =
+      tester.findElement((element) => element is MultiChildRenderObjectElement);
+  expect(element, isNotNull);
+  expect(element.renderObject is RenderStack, isTrue);
+  RenderStack renderObject = element.renderObject;
+  try {
+    RenderObject child = renderObject.firstChild;
+    for (TestParentData expected in expectedParentData) {
+      expect(child is RenderDecoratedBox, isTrue);
+      RenderDecoratedBox decoratedBox = child;
+      expect(decoratedBox.parentData is StackParentData, isTrue);
+      StackParentData parentData = decoratedBox.parentData;
+      expect(parentData.top, equals(expected.top));
+      expect(parentData.right, equals(expected.right));
+      expect(parentData.bottom, equals(expected.bottom));
+      expect(parentData.left, equals(expected.left));
+      child = decoratedBox.parentData.nextSibling;
+    }
+    expect(child, isNull);
+  } catch (e) {
+    print(renderObject.toStringDeep());
+    rethrow;
+  }
+}
+
+final TestParentData kNonPositioned = new TestParentData();
+
+void main() {
+  dynamic cachedException;
+
+  setUp(() {
+    assert(cachedException == null);
+    debugWidgetsExceptionHandler = (String context, dynamic exception, StackTrace stack) {
+      cachedException = exception;
+    };
+  });
+
+  tearDown(() {
```

Comment:
maybe put this at the bottom after the tests so it's more obvious it happens last?